core/src/services/koofr/core.rs

```diff
@@ -0,0 +1,455 @@
+pub struct KoofrCore {
+    /// The root of this core.
+    pub root: String,
+    /// The endpoint of this backend.
+    pub endpoint: String,
+    /// Koofr email
+    pub email: String,
+    /// Koofr password
+    pub password: String,
+
+    /// signer of this backend.
+    pub signer: Arc<RwLock<KoofrSigner>>,
+
+    pub client: HttpClient,
+}
+
+impl Debug for KoofrCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("root", &self.root)
+            .field("endpoint", &self.endpoint)
+            .field("email", &self.email)
+            .finish_non_exhaustive()
+    }
+}
+
+impl KoofrCore {
+    #[inline]
+    pub async fn send(&self, req: Request<AsyncBody>) -> Result<Response<IncomingAsyncBody>> {
+        self.client.send(req).await
+    }
+
+    pub async fn get_auth_data(&self) -> Result<AuthData> {
+        {
+            let signer = self.signer.read().await;
+            if !signer.auth_data.token.is_empty() {
+                return Ok(signer.auth_data.clone());
+            }
+        }
+
+        {
+            let mut signer = self.signer.write().await;
+
+            let url = format!("{}/token", self.endpoint);
+
+            let body = json!({
+                "email": self.email,
+                "password": self.password,
+            });
+
+            let bs = serde_json::to_vec(&body).map_err(new_json_serialize_error)?;
+
+            let req = Request::post(url)
+                .header(header::CONTENT_TYPE, "application/json")
+                .body(AsyncBody::Bytes(Bytes::from(bs)))
+                .map_err(new_request_build_error)?;
+
+            let resp = self.client.send(req).await?;
+
+            let status = resp.status();
+
+            if status != StatusCode::OK {
+                return Err(parse_error(resp).await?);
+            }
+
+            let bs = resp.into_body().bytes().await?;
+            let resp: TokenResponse =
+                serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+            signer.auth_data = AuthData {
+                token: resp.token,
+                mount_id: "".to_string(),
+            };
+
+            let req = Request::get(format!("{}/api/v2/mounts", self.endpoint,));
+
+            let req = req.header(
+                header::AUTHORIZATION,
+                format!("Token token={}", signer.auth_data.token),
+            );
+
+            let req = req
+                .body(AsyncBody::Empty)
+                .map_err(new_request_build_error)?;
+
+            let resp = self.send(req).await?;
+
+            let status = resp.status();
+
+            if status != StatusCode::OK {
+                return Err(parse_error(resp).await?);
+            }
+
+            let bs = resp.into_body().bytes().await?;
+
+            let resp: MountsResponse =
+                serde_json::from_slice(&bs).map_err(new_json_deserialize_error)?;
+
+            for mount in resp.mounts {
```

Comment:
By the way, should `mount` by specifed by user? It seems user can have different mounts at the same time?

Reply:
Can we provide a mount id configuration and use primary mount if the user does not fill it in?

Reply:
> Can we provide a mount id configuration and use primary mount if the user does not fill it in?

This feature seems fine to me, but perhaps we could include it in subsequent PRs to simplify this one.